src/kv/store.h

```diff
@@ -219,35 +219,76 @@ namespace kv
       return *result;
     }
 
-    std::unique_ptr<Snapshot> snapshot(Version v) override
+    void deserialize(std::unique_ptr<Snapshot>& snapshot)
     {
-      std::lock_guard<SpinLock> mguard(maps_lock);
-
-      if (v > current_version())
       {
-        throw ccf::ccf_logic_error(fmt::format(
-          "Attempting to snapshot at invalid version v:{}, current_version:{}",
-          v,
-          current_version()));
+        std::lock_guard<SpinLock> vguard(version_lock);
+        version = snapshot->get_version();
+        last_replicated = snapshot->get_version();
+        last_committable = snapshot->get_version();
       }
 
-      auto snapshot = std::make_unique<Snapshot>();
-
+      std::lock_guard<SpinLock> mguard(maps_lock);
       for (auto& map : maps)
       {
         map.second->lock();
       }
 
-      for (auto& map : maps)
+      auto& snapshots = snapshot->get_snapshots();
+      CCF_ASSERT_FMT(
+        maps.size() == snapshots.size(),
+        "Number of maps does not match the snapshot, maps:{}, snapshots:{}",
+        maps.size(),
+        snapshots.size());
+      for (auto& s : snapshots)
       {
-        snapshot->add_snapshot(std::move(map.second->snapshot(v)));
+        auto search = maps.find(s->get_name());
+        if (search == maps.end())
+        {
+          throw ccf::ccf_logic_error(
+            fmt::format("Map does not exist - {}", s->get_name()));
+        }
+
+        search->second->apply(s);
       }
 
       for (auto& map : maps)
       {
         map.second->unlock();
       }
+    }
 
+    std::unique_ptr<Snapshot> snapshot(Version v) override
+    {
+      auto snapshot = std::make_unique<Snapshot>(v);
+
+      {
+        std::lock_guard<SpinLock> mguard(maps_lock);
+
+        if (v > current_version())
```

Comment:
We also need to check that `v >= commit_version()`.

Reply:
why >= ? Should we not be able to create a snapshot of the current version?

Reply:
`commit_version()` refers to the version at which the map was compacted (i.e. the first element in the roll). We should not be able to snapshot before this version.